lib/improver/standardise.py

```diff
@@ -0,0 +1,446 @@
+        and that obtained by regridding the source grid landmask, to ensure
+        that the "land" or "sea" nature of the points in the regridded cube
+        matches that of the target grid.
+
+        Returns:
+            iris.cube.Cube: Adjusted cube
+        """
+        if self.landmask_name not in self.landmask_source_grid.name():
+            msg = ("Expected {} in input_landmask cube but found {}".format(
+                self.landmask_name, repr(self.landmask_source_grid)))
+            warnings.warn(msg)
+
+        if self.landmask_name not in target_grid.name():
+            msg = ("Expected {} in target_grid cube but found {}".format(
+                self.landmask_name, repr(target_grid)))
+            warnings.warn(msg)
+
+        plugin = AdjustLandSeaPoints(vicinity_radius=self.landmask_vicinity)
+        return plugin.process(cube, self.landmask_source_grid, target_grid)
+
+    def _regrid_to_target(self, cube, target_grid):
+        """
+        Regrid cube to target_grid and inherit appropriate grid attributes
+
+        Returns:
+            iris.cube.Cube: Regridded cube with updated attributes
+        """
+        regridder = Linear(extrapolation_mode=self.extrapolation_mode)
+        if "nearest" in self.regrid_mode:
+            regridder = Nearest(extrapolation_mode=self.extrapolation_mode)
+        cube = cube.regrid(target_grid, regridder)
+
+        if self.REGRID_REQUIRES_LANDMASK[self.regrid_mode]:
+            cube = self._adjust_landsea(cube, target_grid)
+
+        attributes_to_inherit = (
+            {key: val for (key, val) in target_grid.attributes.items()
+             if key in self.grid_attributes})
+        amend_attributes(cube, attributes_to_inherit)
+
+        return cube
+
+    @staticmethod
+    def _standardise_time_coordinates(cube):
+        """
+        If cube time-type coordinates do not conform to expected standards,
+        update units and datatypes in place.
+        """
+        def _update_time_coordinate(coord):
+            """Update a non-conforming time coordinate"""
+            if coord.units.is_time_reference():
+                coord.convert_units(TIME_REFERENCE_UNIT)
+                coord.points = coord.points.astype(TIME_REFERENCE_DTYPE)
+                if coord.bounds is not None:
+                    coord.bounds = coord.bounds.astype(TIME_REFERENCE_DTYPE)
+            else:
+                coord.convert_units(TIME_UNIT)
+                coord.points = coord.points.astype(TIME_DTYPE)
+                if coord.bounds is not None:
+                    coord.bounds = coord.bounds.astype(TIME_DTYPE)
+
+        # check all time coordinates; if check fails, update them all
+        try:
+            check_time_coordinate_metadata(cube)
+        except ValueError:
+            for time_coord in TIME_COORD_NAMES:
+                try:
+                    coord = cube.coord(time_coord)
+                except CoordinateNotFoundError:
+                    pass
```

Comment:
I hadn't considered the use case in which the necessary coordinates might not exist.  If a time-based coordinate is not present for a constant field, then this is fine - we shouldn't be trying to post-process ancillaries.  If it's a field that's constant over a certain time period, then it should have a time coordinate with bounds.

If the input does not have at least two of those coordinates and it needs them, then its metadata are incomplete (and therefore I would argue incorrect, as they are insufficient to describe the data).  I think it's not unreasonable to expect input metadata to be correct and complete.  We can't add metadata where it's not present, and if the user knows their metadata to be incomplete, there are standard tools they can use to correct that problem.

In terms of coordinate names: `time`, `forecast_reference_time` and `forecast_period` are the CF standard names expected for describing this metadata.  So I think we can expect these names on our inputs.  Even if the names were incorrect, we should not be adding essentially duplicate coordinates, but renaming those present.  Again, I suggest that if our documentation is clear on the expectation of CF standards (which it may not be yet), our code doesn't need to take responsibility for converting to those standards.
http://cfconventions.org/Data/cf-standard-names/28/build/cf-standard-name-table.html

Reply:
Yes, I think that's fine, as the metadata shouldn't be incomplete.